packages/search/src/use-fuzzy-search.ts

```diff
@@ -0,0 +1,69 @@
+import Fuse from 'fuse.js';
+import { useState, useMemo, useEffect } from 'react';
+
+const defaultOptions = {
+	threshold: 0.4,
+	distance: 20,
+};
+
+type FieldsProp< T > = T extends string
+	? {
+			fields?: never;
+	  }
+	: {
+			fields: Fuse.IFuseOptions< T >[ 'keys' ];
+	  };
+
+export type UseFuzzySearchOptions< T > = {
+	data: T[];
+	options?: Partial< Fuse.IFuseOptions< T > >;
+	initialQuery?: string;
+} & FieldsProp< T >;
+
+export const useFuzzySearch = < T >( {
+	data,
+	fields = [],
+	initialQuery = '',
+	options = defaultOptions,
+}: UseFuzzySearchOptions< T > ) => {
+	const [ query, setQuery ] = useState( initialQuery );
+
+	/**
+	 * We want to re-use the `Fuse` instance because creating one every time
+	 * the dataset changes is an expensive operation.
+	 *
+	 * I first used `useRef` to save that instance, but that hook does not allow
+	 * lazy instantiation of the initial value, so I had to create wrapper functions
+	 * and the result didn't look great.
+	 *
+	 * Falling back to `useState` fixes that problem, though now we have to add
+	 * the `fuseInstance` variable to the dependency array.
+	 *
+	 */
+	const [ fuseInstance ] = useState< Fuse< T > >( () => {
+		return new Fuse< T >( data, {
```

Comment:
The explicit generic parameters aren't needed here and it might be better to leave them out. It allows the libraries to change how their type definitions work, and as long as they continue to support type inference, our code will just keep working.

```suggestion
	const [ fuseInstance ] = useState( () => {
		return new Fuse( data, {
```